HotPathAllocationAnalyzer/Analyzers/ExplicitAllocationAnalyzer.cs

```diff
@@ -43,32 +45,54 @@ public ExplicitAllocationAnalyzer()
             var node = context.Node;
             var semanticModel = context.SemanticModel;
-            Action<Diagnostic> reportDiagnostic = context.ReportDiagnostic;
+            var reportDiagnostic = context.ReportDiagnostic;
             var cancellationToken = context.CancellationToken;
-            string filePath = node.SyntaxTree.FilePath;
-
-            // An InitializerExpressionSyntax has an ObjectCreationExpressionSyntax as it's parent, i.e
-            // var testing = new TestClass { Name = "Bob" };
-            //               |             |--------------| <- InitializerExpressionSyntax or SyntaxKind.ObjectInitializerExpression
-            //               |----------------------------| <- ObjectCreationExpressionSyntax or SyntaxKind.ObjectCreationExpression
-            var initializerExpression = node as InitializerExpressionSyntax;
-            if (initializerExpression?.Parent is ObjectCreationExpressionSyntax objectCreation)
+            var filePath = node.SyntaxTree.FilePath;
+
+            if (node is ObjectCreationExpressionSyntax newObj)
             {
-                var typeInfo = semanticModel.GetTypeInfo(objectCreation, cancellationToken);
-                if (typeInfo.ConvertedType?.TypeKind != TypeKind.Error &&
-                    typeInfo.ConvertedType?.IsReferenceType == true &&
-                    objectCreation.Parent?.IsKind(SyntaxKind.EqualsValueClause) == true &&
-                    objectCreation.Parent?.Parent?.IsKind(SyntaxKind.VariableDeclarator) == true)
+                var typeInfo = semanticModel.GetTypeInfo(newObj, cancellationToken);
+                if (typeInfo.ConvertedType != null && typeInfo.ConvertedType.TypeKind != TypeKind.Error && typeInfo.ConvertedType.IsReferenceType)
+                {
+                    var (ancestorType, ancestor) = newObj.FindAncestor(SyntaxKind.SimpleAssignmentExpression,
+                                                                       SyntaxKind.ComplexElementInitializerExpression,
+                                                                       SyntaxKind.SimpleAssignmentExpression,
+                                                                       SyntaxKind.VariableDeclaration,
+                                                                       SyntaxKind.InvocationExpression);
+                    reportDiagnostic(ancestor == null
+                                         ? Diagnostic.Create(NewObjectRule, newObj.GetLocation(), EmptyMessageArgs)
+                                         : Diagnostic.Create(NewObjectRule, ancestor.GetLocation(), EmptyMessageArgs));
+
+                    HeapAllocationAnalyzerEventSource.Logger.NewObjectCreationExpression(filePath);
+                }
+
+                return;
+            }
+
+            if (node is InitializerExpressionSyntax objectInitializerSyntax)
+            {
+                if (node.Kind() != SyntaxKind.ObjectInitializerExpression)
+                    return;
```

Comment:
Is that needed ? How is that triggered ?

Reply:
We probably don't want to report an allocation for a collection initialization or ComplexElementInitializerExpression
since we alreay report it for the collection explicit creation.
For exemple in : 
Dictionary<int, List<int>> data = new() 
{  //CollectionInitializerExpression
   { 1, new() { 1, 3, 5 } }, //ComplexElementInitializerExpression
}
It would be annoying to have 1 report for every element of the dictionary
If you don't do the node check the ComplexElementInitializerExpression will find an objectCreation ancestor thus reporting.